core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1264,6 +1270,7 @@ class DAGScheduler(
             try { // cancelTasks will fail if a SchedulerBackend does not implement killTask
               taskScheduler.cancelTasks(stageId, shouldInterruptThread)
               stage.latestInfo.stageFailed(failureReason)
+              outputCommitCoordinator.stageEnd(stage.id)
```

Comment:
It actually looks like `handleTaskCompletion` has a nested `markStageAsFinished` method that looks like it should do this. That method also removes the stage from `runningStages`, which doesn't appear to happen in all of the paths where we post `SparkListenerStageCompleted`.  Let me take a look at this and see whether there's a safe way to refactor things to use this method.